tests/unit/document_tests.py

```diff
@@ -622,6 +622,47 @@ def test_document_context_manager_no_doc_id(self):
         self.assertTrue(doc['_rev'].startswith('1-'))
         self.assertEqual(self.db['julia006'], doc)
 
+    def test_document_context_manager_creation_failure_on_error(self):
+        """
+        Test that the document context manager skips document creation if there
+        is an error.
+        """
+        with self.assertRaises(ZeroDivisionError), Document(self.db, 'julia006') as doc:
+            doc['name'] = 'julia'
+            doc['age'] = 6
+            raise ZeroDivisionError()
+
+        doc = Document(self.db, 'julia006')
+        try:
+            doc.fetch()
+        except requests.HTTPError as err:
+            self.assertEqual(err.response.status_code, 404)
+        else:
+            self.fail('Above statement should raise a HTTPError.')
+
+    def test_document_context_manager_update_failure_on_error(self):
+        """
+        Test that the document context manager skips document update if there
+        is an error.
+        """
+        # Create the document.
+        doc = Document(self.db, 'julia006')
+        doc['name'] = 'julia'
+        doc['age'] = 6
+        doc.save()
+
+        # Make a document update and then raise an error.
+        with self.assertRaises(ZeroDivisionError), Document(self.db, 'julia006') as doc:
+            doc['age'] = 7
+            raise ZeroDivisionError()
+
+        # Assert the change persists locally.
+        self.assertEqual(doc['age'], 7)
```

Comment:
Note, if we raise an exception in the block then the local copy still gets updated. It's only the remote save that gets skipped.

Reply:
I wonder if there's a way for us to use a temporary dict during the context manager that only updates the local doc and save remotely if there's no exception.
If not, we should add a note in the context manager docs that if an exception occurs, the remote doc is not saved but the local doc is.

Reply:
I'll add a further note in the docs. I think a temp dict could add too many complications.